### ISSUE TITLE
panend fires before swipe events do...

### ISSUE THREAD

**Issue by glomotion** [hammerjs/hammer.js#882]
This may not be an issue with Hammer, and also may not be the right place to ask this question, so apologies in advance if this is the case. :)

I've built my own carousel on hammer 2.0 - much like the original hammer carousel (here: http://codepen.io/berkin/pen/HKgnF).

Put simply, I'm having trouble with the panend and swipe events, and how they play together. 

I can recognise both Swipe and Pan events just fine (using recogniseWith) - but my problem is swipe events fire after the panend event - which forces me to wrap code in "panend" inside a setTimeout( (),5 ) - to ensure that Swipe events do in fact run before the panend. 

Just FYI: the usecase is pretty simple > On "panend" i check how far the user has panned - and if they've panned enough, trigger code to change the page. Ontop of this I use up and down swipes to bypass this logic and just go straight to page change. ATM a swipe is causing a page change, and then the code in my 'panend' is firing, possibly detecting that distance-panned is enough, which triggers a second page change - resulting in a single gesture double page change.

I have tried including requireFailure(mc.get('swipe')) on the Pan recogniser, but this simply causes the swipe recogniser to no longer detect swipes. 

eg: 
```javascript
mc = new Hammer.Manager(wrapper);

mc.add(new Hammer.Pan({
  direction: Hammer.DIRECTION_VERTICAL,
  threshold: 5,
  pointers: 0,
  touchAction: 'pan-x'
}));

mc.add(new Hammer.Swipe({
  direction: Hammer.DIRECTION_VERTICAL,
  threshold: 40,
  velocity: 0.65,
  touchAction: 'pan-x'
})).recognizeWith(mc.get('pan'));

mc.get('pan').requireFailure(mc.get('swipe'));
```

Any help you guys could give me around what I'm doing wrong / how this could work better, would be hugely appreciated! :)

**Comment by josephmcasey** [comment]
Have you tried changing the threshold of your Swipe to 5 or less.

Otherwise, can you provide a working code zip or repo?

**Comment by BAM5** [comment]
Try adding Swipe to the manager before Pan?

**Comment by glomotion** [comment]
Ohhh nice. Thanks guys! 

@BAM5: I never imagined the order of how they get added to the manager would affect the event fire timing. Flipping the order fixed everything. Thanks again! :)

**Comment by BAM5** [comment]
Yep, each recognizer gets called upon to process input data in the order you add it. So in your initial post it's only natural that the Pan recognizer would get processed first and would emit its event first. I've done a bit of digging through the source and have made my own recognizer so it may not be as intuitive as I find it :)

**Comment by glomotion** [comment]
Ahhh damn. Looks like i jumped the gun / didn't test properly. 

No matter how i order the recognisers, I can't seem to change the panend / swipe event order.

Ported my current code into codepen here:
http://s.codepen.io/glomotion/debug/EPaYoq

**Comment by BAM5** [comment]
You sure? I'm getting swipe events before pan events on chrome.

**Comment by glomotion** [comment]
So actually, I think my problem is actually a little more complicated than I first thought. 

It's not the order thats the problem I don't think - but the fact that they both pretty much run together asynchronously. 

I need to be able to affect the panend event as soon as i detect a swipe event. But it looks like they both fire just about simultaneously. Eg. I set a var in the on swipe event - but by the time panend fires, this var setting doesn't seem to have happened yet. 

Which means logic inside one can't seem to affect the other. Does this make any sense?

**Comment by BAM5** [comment]
Erm, if the swipe event is firing before the pan event, you should have no problem setting a flag from the swipe listener that the panend listener can see. Perhaps the flag variable you're using isn't shared among the listeners' scopes? Try setting a global on window to test.

**Comment by glomotion** [comment]
Interesting. Wasn't a scope issue, but a variable setting issue coming from the panup/pandown. Think I've got it now. Thanks again for your help. :)

**Comment by BAM5** [comment]
No prob! Good luck!

**Comment by runspired** [comment]
There is also a bug with `panend` being fired before the last `pan<Dir>` event which has been fixed on master and is possibly contributing to this.

**Comment by glomotion** [comment]
Ohhh true. Thanks Chris! Is that fix you mention part of hammer 2.0.4 - or was it added more recently than that?

**Comment by arschmitz** [comment]
its more recent it will be in the 2.05 release we are trying to get out ASAP
